proposal/new-js-importer.d.ts

```diff
@@ -0,0 +1,322 @@
+ * in-memory, the importer should choose a custom URL scheme to guarantee that
+ * its canonical URLs don't conflict with any other importer's.
+ *
+ * Having a canonical URL for each stylesheet allows Sass to ensure that the
+ * same stylesheet isn't loaded multiple times in the new module system.
+ *
+ * ### Loading
+ *
+ * The second step actually loads the text of the stylesheet. The `load()`
+ * method takes a canonical URL that was returned by `canonicalize()` and
+ * returns the contents of the stylesheet at that URL. This is only called once
+ * per compilation for each canonical URL; future loads of the same URL will
+ * re-use either the existing module (for `@use` and `@forward`) or the parse
+ * tree (for `@import`).
+ *
+ * ### `FileImporter`
+ *
+ * This proposal also adds a special type of importer known as a `FileImporter`.
+ * This importer makes the common case of redirecting loads to somewhere on the
+ * physical filesystem easier. It doesn't require the caller to implement
+ * `load()`, since that's always going to be the same for files on disk.
+ *
+ * A `FileImporter` only needs to have one method: `findFileUrl()`, which takes
+ * a relative URL and returns the absolute `file:` URL it should refer to.
+ * However, this URL doesn't need to be fully canonicalized: the Sass compiler
+ * will take care of resolving partials, file extensions, index files, and so
+ * on.
+ */
+
+/** ## API */
+
+import {URL} from 'url';
+
+import {Options, Syntax} from './new-js-api';
+import './new-js-api';
+
+declare module './new-js-api' {
+  interface Options<sync extends 'sync' | 'async'> {
+    importers?: _Importer<sync>[];
+  }
+}
+
+/** This definition supersedes the one in the New JavaScript API proposal. */
+type StringOptions<sync extends 'sync' | 'async'> = Options<sync> & {
+  syntax?: Syntax;
+} & (
+    | {url?: URL}
+    | {
+        importer: _Importer<sync>;
+        url: URL;
+      }
+  );
+
+/**
+ * This interface represents an [importer]. When the importer is invoked with a
+ * string `string`:
+ *
+ * [importer]: ../spec/modules.md#importer
+ *
+ * * Let `fromImport` be `true` if the importer is being run for an `@import`
+ *   and `false` otherwise.
+ *
+ * * Let `url` be the result of calling `findFileUrl` with `url` and
+ *   `fromImport`.
+ *
+ * * If `url` is null, return null.
+ *
+ * * If `url`'s scheme is not `file`, throw an error.
+ *
+ * * Let `resolved` be the result of [resolving `url`].
+ *
+ * * If `resolved` is null, return null.
+ *
+ * * Let `text` be the contents of the file at `resolved`.
+ *
+ * * Let `syntax` be:
+ *   * "scss" if `url` ends in `.scss`.
+ *   * "indented" if `url` ends in `.sass`.
+ *   * "css" if `url` ends in `.css`.
+ *
+ *   > The algorithm for resolving a `file:` URL guarantees that `url` will have
+ *   > one of these extensions.
+ *
+ * * Return `text`, `syntax`, and `resolved`.
+ *
+ * [resolving `url`]: ../spec/modules.md#resolving-a-file-url
+ */
```

Comment:
Consider speccing how `sourceMapUrl` is constructed differently from `url`.

Reply:
As above, this isn't actually a JS `Importer`.